### PR TITLE
🛣️ feat: Add MCP Remote Proxy Support

### LINKED ISSUE #12995: [Enhancement]: Support outbound proxy configuration for remote MCP HTTP/SSE transports
### What happened?

I'm trying to connect a remote Splunk MCP server from LibreChat through a proxy, but I'm getting a connection timeout error.

The MCP endpoint is reachable from the API container using curl, so network connectivity appears to be working correctly. Since we are using a proxy, we added the proxy-related headers/configuration in the YAML request settings, and the connection is configured like this:

```
splunk:
    startup: false
    type: streamable-http
    url: <splunk_url>
    iconPath: /assets/splunk.png
    timeout: 80000
    initTimeout: 30000
    headers:
      Authorization: "Bearer {{splunk_api_key}}"
      Content-Type: "application/json"
      HTTPS_PROXY: <proxy_address>
      Accept: "application/json, text/event-stream"
    customUserVars:
      splunk_api_key:
        title: "Splunk API Key"
        description: "Enter your Splunk Bearer token"
```
We are using curl like this from the LibreChat container, and the connection is successful.
```
curl -v -X POST \
  --proxy <proxy_address>/ \
  -H "Authorization: Bearer <token>" \
  -H "Content-Type: application/json" \
  -H "Accept: application/json, text/event-stream" \
  -d '{"jsonrpc":"2.0","method":"initialize","params":{"protocolVersion":"2024-11-05","capabilities":{}},"id":1}' \
<splunk_url>
```

Could you please check this

### Version Information

1bc2692a1564af381889ab4d83f1d305d88552e9

### Steps to Reproduce

1. Connect the remote Splunk MCP server from LibreChat through a proxy.

### What browsers are you seeing the problem on?

_No response_

### Relevant log output

```shell
2026-05-07 21:06:13 info: [updateUserPluginsController] Attempting disconnect of MCP server "splunk" for user 6862786e99cd9d0127a61df8 after plugin auth update.
2026-05-07 21:06:13 info: [MCP Reinitialize] Reinitializing server: splunk
2026-05-07 21:06:13 info: [MCP Reinitialize] Disconnected existing user connection for server: splunk
2026-05-07 21:06:13 info: [MCP][User: 6862786e99cd9d0127a61df8][splunk] Establishing new connection
2026-05-07 21:06:14 info: [MCP][User: 6862786e99cd9d0127a61df8][splunk] Access token missing
2026-05-07 21:06:14 info: [MCP][User: 6862786e99cd9d0127a61df8][splunk] Access token missing and no refresh token available — re-authentication required
2026-05-07 21:06:14 info: [MCP][splunk][6862786e99cd9d0127a61df8] Re-authentication required for "splunk": access token missing and no refresh token available, will trigger OAuth flow
2026-05-07 21:06:14 info: [MCP][User: 6862786e99cd9d0127a61df8][splunk] Creating streamable-http transport: <url>
2026-05-07 21:06:24 error: [MCP][User: 6862786e99cd9d0127a61df8][splunk] Transport error (transient, will reconnect): fetch failed (request aborted, likely after a timeout — connection may still be usable)
2026-05-07 21:06:24 info: [MCP][User: 6862786e99cd9d0127a61df8][splunk] Streamable-http transport closed
2026-05-07 21:06:24 error: [MCP][User: 6862786e99cd9d0127a61df8][splunk] Connection failed: fetch failed
2026-05-07 21:06:26 info: [MCP][User: 6862786e99cd9d0127a61df8][splunk] Creating streamable-http transport: <url>
2026-05-07 21:06:37 error: [MCP][User: 6862786e99cd9d0127a61df8][splunk] Transport error (transient, will reconnect): fetch failed (request aborted, likely after a timeout — connection may still be usable)
2026-05-07 21:06:37 info: [MCP][User: 6862786e99cd9d0127a61df8][splunk] Streamable-http transport closed
2026-05-07 21:06:37 error: [MCP][User: 6862786e99cd9d0127a61df8][splunk] Connection failed: fetch failed
2026-05-07 21:06:41 info: [MCP][User: 6862786e99cd9d0127a61df8][splunk] Creating streamable-http transport: <url>
2026-05-07 21:06:44 error: [MCP][User: 6862786e99cd9d0127a61df8][splunk] Failed to establish connection Connection timeout after 30000ms
2026-05-07 21:06:44 info: [MCP Reinitialize] getConnection threw error: Connection timeout after 30000ms
2026-05-07 21:06:44 info: [MCP Reinitialize] OAuth state - oauthRequired: false, oauthUrl: null
2026-05-07 21:06:44 error: [MCP Reinitialize] Error initializing MCP server splunk for user: Connection timeout after 30000ms
```

### Screenshots

None

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by danny-avila**
## Summary

I added outbound proxy support for remote MCP SSE and Streamable HTTP transports, resolving #12995.

- Added an admin-configurable `proxy` field for remote MCP transports with URL/protocol validation and environment variable resolution.
- Honored LibreChat's `PROXY` env var, plus standard `HTTP_PROXY`, `HTTPS_PROXY`, and `NO_PROXY` fallbacks when no per-server proxy is configured.
- Routed MCP SSE and Streamable HTTP fetch dispatchers through undici `ProxyAgent` while preserving separate POST and long-lived SSE GET timeouts.
- Kept proxy configuration out of user-created MCP server input so end users cannot choose outbound proxy targets.
- Added proxied-target SSRF preflight checks before network dispatch and preserved redirect credential stripping, safe redirect dispatchers, and `NO_PROXY` direct-connect SSRF guards.
- Documented the new per-server YAML field and clarified the env proxy behavior in `.env.example`.

## Change Type

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update
- [x] Documentation update

## Testing

- Ran `npm run smart-reinstall` from the repo root.
- Ran `npx turbo run build --filter=librechat-data-provider --filter=@librechat/api` from the repo root.
- Ran `npx jest specs/mcp.spec.ts --runInBand --coverage=false` from `packages/data-provider`.
- Ran `npx jest src/mcp/__tests__/mcp.spec.ts src/utils/env.spec.ts src/mcp/__tests__/MCPConnectionSSRF.test.ts --runInBand --coverage=false` from `packages/api`.
- Ran `npx eslint packages/api/src/mcp/connection.ts packages/api/src/mcp/__tests__/MCPConnectionSSRF.test.ts packages/api/src/mcp/__tests__/mcp.spec.ts packages/api/src/utils/env.ts packages/api/src/utils/env.spec.ts packages/data-provider/src/mcp.ts packages/data-provider/specs/mcp.spec.ts` from the repo root.
- Ran `git diff --check` from the repo root.

### **Test Configuration**:

- Node.js: `v20.19.5`
- npm: `10.8.2`
- Base branch: `dev` (`36e95353e`)

## Checklist

- [x] My code adheres to this project's style guidelines
- [x] I have performed a self-review of my own code
- [x] I have commented in any complex areas of my code
- [x] I have made pertinent documentation changes
- [x] My changes do not introduce new warnings
- [x] I have written tests demonstrating that my changes are effective or that my feature works
- [x] Local unit tests pass with my changes

### CODE REVIEW

**Comment by danny-avila** [comment]
@codex review

**Comment by inv-Eldho** [comment]
@danny-avila 
Has the development been completed? Could you please provide an update on this? If it’s completed, please merge it into the main branch.

File: packages/api/src/mcp/connection.ts [hunk 1]
```diff
@@ -196,6 +210,166 @@ function getUrlPort(url: URL | string): string {
...
+    type: 'env',
+    httpProxy,
+    httpsProxy,
+    noProxy: getTrimmedEnv('no_proxy', 'NO_PROXY'),
+  };
+}
+
+function getProxyEntryPort(entry: string): { hostname: string; port: number } {
+  const parsed = entry.match(/^(.+):(\d+)$/);
+  return {
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/mcp/connection.ts]: Fixed in 695935ef6. `NO_PROXY` parsing now treats a port suffix as a port only for bracketed IPv6 entries like `[::1]:3000` or single-colon `hostname:port` entries, so bare IPv6 literals such as `::1` and `2001:db8::1` remain host matches. Added regression coverage in `packages/api/src/mcp/__tests__/MCPConnectionSSRF.test.ts`.

Verification:
- `npm run smart-reinstall`
- `npm run build` in `packages/data-schemas`
- `npm run build` in `packages/data-provider`
- `npx jest src/utils/env.spec.ts --runInBand`
- `npx jest src/mcp/__tests__/MCPConnectionSSRF.test.ts --runInBand`
- `npx eslint packages/api/src/utils/env.ts packages/api/src/utils/env.spec.ts packages/api/src/mcp/connection.ts packages/api/src/mcp/__tests__/MCPConnectionSSRF.test.ts`

File: packages/api/src/mcp/connection.ts [hunk 2]
```diff
@@ -196,6 +210,166 @@ function getUrlPort(url: URL | string): string {
...
+    hostname: (parsed ? parsed[1] : entry)
+      .replace(/^\*?\./, '')
+      .replace(/^\[|\]$/g, '')
+      .toLowerCase(),
+    port: parsed ? Number.parseInt(parsed[2], 10) : 0,
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/mcp/connection.ts]: Fixed in 695935ef6. `NO_PROXY` parsing now treats a port suffix as a port only for bracketed IPv6 entries like `[::1]:3000` or single-colon `hostname:port` entries, so bare IPv6 literals such as `::1` and `2001:db8::1` remain host matches. Added regression coverage in `packages/api/src/mcp/__tests__/MCPConnectionSSRF.test.ts`.

Verification:
- `npm run smart-reinstall`
- `npm run build` in `packages/data-schemas`
- `npm run build` in `packages/data-provider`
- `npx jest src/utils/env.spec.ts --runInBand`
- `npx jest src/mcp/__tests__/MCPConnectionSSRF.test.ts --runInBand`
- `npx eslint packages/api/src/utils/env.ts packages/api/src/utils/env.spec.ts packages/api/src/mcp/connection.ts packages/api/src/mcp/__tests__/MCPConnectionSSRF.test.ts`

**Comment by danny-avila** [comment]
@codex review

**Comment by danny-avila** [comment]
@codex review

File: packages/api/src/mcp/connection.ts [hunk 3]
```diff
@@ -541,6 +552,176 @@ function getUrlPort(url: URL | string): string {
...
+    return false;
+  }
+
+  const trimmed = noProxy.trim();
+  if (!trimmed) {
+    return false;
+  }
+  if (trimmed === '*') {
+    return true;
+  }
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/mcp/connection.ts:775]: Fixed in 87ddb8596. `shouldBypassEnvProxy` now treats `*` as a global bypass token even when it appears inside a comma/space-separated `NO_PROXY` list such as `localhost,*`. Added regression coverage in `packages/api/src/mcp/__tests__/MCPConnectionSSRF.test.ts`.

Verification:
- `npx jest src/mcp/__tests__/MCPConnectionSSRF.test.ts --runInBand`
- `npx eslint packages/api/src/mcp/connection.ts packages/api/src/mcp/__tests__/MCPConnectionSSRF.test.ts`
- `npm run build` in `packages/api`

**Comment by danny-avila** [comment]
@codex review

File: packages/api/src/mcp/connection.ts [hunk 4]
```diff
@@ -926,58 +1112,70 @@ export class MCPConnection extends EventEmitter {
     baseUrl?: string,
     guardStreamableHTTPResponses = false,
   ): (input: UndiciRequestInfo, init?: UndiciRequestInit) => Promise<UndiciResponse> {
+    const proxyConfig = this.proxyConfig;
+    const initialProxyUrl = baseUrl ? getProxyUrlForRequest(proxyConfig, baseUrl) : undefined;
     const basePort = baseUrl ? getUrlPort(baseUrl) : '';
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/mcp/connection.ts]: Fixed in 146be95c6. `customFetch` now chooses and caches dispatchers from each resolved request URL instead of the initial MCP base URL, so absolute message endpoints get their own `NO_PROXY`/proxy decision. Added a regression that routes a base proxied server to a `127.0.0.1` `NO_PROXY` request directly.

Verification:
- `npx jest src/mcp/__tests__/MCPConnectionSSRF.test.ts --runInBand`
- `npx eslint packages/api/src/mcp/connection.ts packages/api/src/mcp/__tests__/MCPConnectionSSRF.test.ts`
- `npm run build` in `packages/api`

File: packages/api/src/mcp/connection.ts [hunk 5]
```diff
@@ -541,6 +552,179 @@ function getUrlPort(url: URL | string): string {
...
+): Promise<void> {
+  if (!proxyConfig || !useSSRFProtection) {
+    return;
+  }
+
+  const targetUrl = new URL(urlString);
+  const port = getUrlPort(targetUrl);
+  const isBlockedTarget =
+    isSSRFTarget(targetUrl.hostname, allowedAddresses, port) ||
+    (await resolveHostnameSSRF(targetUrl.hostname, allowedAddresses, port));
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/mcp/connection.ts:890]: Fixed in 146be95c6. Proxied SSRF preflight now fails closed when the app host cannot resolve the target hostname before handing the request to a proxy, so proxy-only DNS names cannot bypass the check. Added regression coverage for the DNS failure path.

Verification:
- `npx jest src/mcp/__tests__/MCPConnectionSSRF.test.ts --runInBand`
- `npx eslint packages/api/src/mcp/connection.ts packages/api/src/mcp/__tests__/MCPConnectionSSRF.test.ts`
- `npm run build` in `packages/api`

**Comment by danny-avila** [comment]
@codex review

File: packages/api/src/mcp/connection.ts [hunk 6]
```diff
@@ -541,6 +561,349 @@ function getUrlPort(url: URL | string): string {
...
+      continue;
+    }
+    if (matchesNoProxyIPPattern(hostname, proxyEntry.hostname)) {
+      return true;
+    }
+    if (proxyEntry.matchSubdomains) {
+      if (hostname.endsWith(`.${proxyEntry.hostname}`)) {
+        return true;
+      }
+      continue;
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/mcp/connection.ts]: Fixed in b5caa0494. `NO_PROXY` host entries now follow Undici env-proxy matching: bare, leading-dot, and `*.` entries normalize to the domain and bypass both the apex host and subdomains, while the dot-boundary check keeps unrelated hosts like `badexample.com` proxied. Added regression coverage for apex/subdomain/boundary cases.

Verification:
- `npx jest src/mcp/__tests__/MCPConnectionSSRF.test.ts --runInBand`
- `npx eslint packages/api/src/mcp/connection.ts packages/api/src/mcp/__tests__/MCPConnectionSSRF.test.ts`
- `npm run build` in `packages/api`

File: packages/api/src/mcp/connection.ts [hunk 7]
```diff
@@ -541,6 +561,349 @@ function getUrlPort(url: URL | string): string {
...
+
+  const targetUrl = new URL(urlString);
+  const port = getUrlPort(targetUrl);
+  const isBlockedTarget =
+    isSSRFTarget(targetUrl.hostname, allowedAddresses, port) ||
+    (await resolveHostnameSSRF(targetUrl.hostname, allowedAddresses, port));
+
+  if (!isBlockedTarget) {
+    await assertProxiedRequestTargetResolvable(targetUrl.hostname);
+    return;
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/mcp/connection.ts:894]: Fixed in b5caa0494. `assertProxiedRequestTargetAllowed` now skips the local DNS resolvability preflight when the proxied target host:port is explicitly listed in `allowedAddresses`, so admin-allowlisted split-DNS/proxy-only names are not rejected before the proxy can resolve them. Added a tunnel-proxy regression that proves no local DNS preflight runs for an allowed `proxy-only.internal:80` target.

Verification:
- `npx jest src/mcp/__tests__/MCPConnectionSSRF.test.ts --runInBand`
- `npx eslint packages/api/src/mcp/connection.ts packages/api/src/mcp/__tests__/MCPConnectionSSRF.test.ts`
- `npm run build` in `packages/api`

File: packages/api/src/mcp/connection.ts [hunk 8]
```diff
@@ -541,6 +561,349 @@ function getUrlPort(url: URL | string): string {
...
+    if (matchesNoProxyIPPattern(hostname, proxyEntry.hostname)) {
+      return true;
+    }
+    if (proxyEntry.matchSubdomains) {
+      if (hostname.endsWith(`.${proxyEntry.hostname}`)) {
+        return true;
+      }
+      continue;
+    }
+    if (hostname === proxyEntry.hostname) {
```
[the remark this answers is not in the document]
Comment (danny-avila) [reply packages/api/src/mcp/connection.ts]: Fixed in b5caa0494. Wildcard and leading-dot `NO_PROXY` entries now share Undici-style normalization, so `*.example.com` and `.example.com` bypass the proxy for both `example.com` and its subdomains instead of missing the apex host. Added regression coverage for wildcard and leading-dot apex/subdomain behavior.

Verification:
- `npx jest src/mcp/__tests__/MCPConnectionSSRF.test.ts --runInBand`
- `npx eslint packages/api/src/mcp/connection.ts packages/api/src/mcp/__tests__/MCPConnectionSSRF.test.ts`
- `npm run build` in `packages/api`

**Comment by danny-avila** [comment]
@codex review

**Comment by danny-avila** [comment]
@codex review
